packages/sprotty/src/features/edge-junction/junction-finder.ts

```diff
@@ -0,0 +1,229 @@
+/********************************************************************************
+ * Copyright (c) 2021-2022 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from "inversify";
+import { SParentElementImpl } from "../../base/model/smodel";
+import { SEdgeImpl } from "../../graph/sgraph";
+import { EdgeRouting, IEdgeRoutePostprocessor, RoutedPoint } from "../routing/routing";
+
+/**
+ * Finds junction points in the edge routes. A junction point is a point where two or more edges split.
+ * This excludes the source and target points of the edges.
+ *
+ * Only works with straight line segments.
+ */
+@injectable()
+export class JunctionFinder implements IEdgeRoutePostprocessor {
+    apply(routing: EdgeRouting, parent: SParentElementImpl): void {
+        this.findJunctions(routing, parent);
+    }
+
+    findJunctions(routing: EdgeRouting, parent: SParentElementImpl) {
+        // gather all edges from the parent
+        const edges = parent.children.filter(child => child instanceof SEdgeImpl) as SEdgeImpl[];
+
+        routing.routes.forEach((route, routeId) => {
+            // for each route we find the corresponding edge from the model by matching the route id and the edge id
+            const edge = edges.find(e => e.id === routeId);
+
+            if (!edge) {
+                return;
+            }
+
+            // we find all edges with the same source as the current edge, excluding the current edge
+            const edgesWithSameSource = edges.filter(e => e.sourceId === edge.sourceId && e.id !== edge.id);
```

Comment:
The number of routes should be more or less the same as the number of edges. This means we have quadratic runtime complexity here: with 1.000 edges, you need 1.000.000 comparisons. One way to reduce the complexity is to create multiple Maps (by `id`, `sourceId` and `targetId`) instead of a single array. However, this would likely degrade the performance for smaller graph. An optimized implementation might choose an algorithm depending on the number of routes, but I wouldn't go that far for now.